.changeset/ten-rats-worry.md

```diff
@@ -0,0 +1,30 @@
+---
+'xstate': minor
+---
+
+You can now strongly type `meta` on state nodes and transitions via `stateMeta` and `transitionMeta` respectively:
+
+```ts
+import { createMachine } from 'xstate';
+
+const machine = createMachine({
+  types: {
+    stateMeta: {} as { title: string },
```

Comment:
I always imagined this to be a per-state information and not a global one. Here, you are going to restrict all state metas to have the same shape. This definitely simplifies some things but it also feels like it cripples the utility of this type.

I think that before introducing any changes to this it would be great to establish where this information pops up in our runtime and types (I wonder if we could accurately list all of those places without actually rechecking the current implementation 😉 ). Having a list like that would allow us to assess the idea of this being a single shape for all states.

Reply:
I think per-state is more of a lesser user-case than a common use-case. And even so, we can use (discriminated) unions for the `stateMeta` value.

Please see here for a motivating example: https://github.com/statelyai/xstate/pull/4239/files#diff-a43cf6af3000f32addb10c6176a04f8a5066bc04358b4805726165c65d8565d3R64

Reply:
This is still a very advanced use case - at that point, I think it should be rather OK to require this metadata to be casted~/asserted at some point in the user's code. The current implementation doesn't even *require* those metas to be provided for each state/transition (and that would complicate types even more so I would prefer not doing that). So there is no strict guarantee that each state/transition will have this information and the user is left alone when it comes to remembering that they need to add this. If they need to add this then requiring smth like `meta: {} satisfies MyShape` doesn't sound like a big ask to me.

I understand the appeal of such generated UIs etc but I'm just not sure if this narrow use case should motivate extra generics here. The `statesMeta` is at least accessible on `state.meta` today (although in an awkward way) so that could be useful at typing this. The `transitionMeta` isn't as publicly available though.

Reply:
See #809 for more motivation